### ISSUE TITLE
Allowing 'watch' clients to request watch bookmarks (or optionally increasing frequency of bookmarks)

### ISSUE THREAD

**Issue by munnerz** [kubernetes/kubernetes#127693]
### What would you like to be added?

It should be possible for a client of the apiserver that is watching a resource type to 'request' a watch bookmark be sent immediately, similar to how etcd allows requesting progress of a watch (https://github.com/etcd-io/etcd/issues/9855).

Alternatively, having some kind of configurable 'higher frequency' (than the current default 1 minute) for bookmarks may be sufficient. This configuration wouldn't be needed for 99% of watch requests, so I'd propose exposing something like `?watchBookmarkFrequency=100ms&allowWatchBookmarks=true`. We'd need to expose some kind of configurable 'floor' for this value, and also look at whether we need to authorise its usage somehow too (once an analysis on scalability is completed and we've got an idea of the relative extra cost of higher-frequency bookmarks).

### Why is this needed?

When building Kubernetes apiserver proxies, to support consistent reads being served from a proxy layer that maintains its own in-memory cache (i.e. another instantiation of cacher.Cacher), we need some way to either request on demand a watch bookmark, or otherwise 'bound' the amount of time the apiserver must wait to some semi-predictable frequency.

This would allow a caching proxy to serve consistent reads from its own cache, with similar assurances to what the WatchList feature provides today (e.g. "we can serve a consistent read within 100ms+(overhead)").

**Comment by munnerz** [comment]
/sig api-machinery

🙃 every. single. time.

**Comment by munnerz** [comment]
... @deads2k @jpbetz too as SIG technical leads :)

**Comment by jpbetz** [comment]
cc @serathius for SIG etcd visibility

I'm also hesitant to expose this aspect of the apiserver/etcd interface to clients.

**Comment by munnerz** [comment]
> Hmm, I'm not sure I'd want to let a client force the API server to opt-into sub-second bookmark requests to etcd.

I agree, this is certainly something that could cause an uptick in traffic per type. I think some form of distinct authorisation should be configured to make up for this. But with sufficient control around *who* can subscribe to this kind of progress notification, it doesn't seem too different to a client watching a particular busy API type (which fires sub-second watch events). Certainly though this is a more privileged operation than a regular watch.

> I'm also hesitant to expose this aspect of the apiserver/etcd interface to clients.

I _did_ consider exposing RequestWatchProgress directly, however it then means we need 'identity' for watches and clever request routing (for the 'request watch progress' endpoint to be routed to the apiserver peer serving a particular watch).

This is what led to the idea of allowing clients to specify a frequency with which they want to receive progress notifications, as this avoids the need to actually expose an on-demand request as a part of our own API/exposing it through. This approach is effectively permitting per-watch tuning of https://github.com/kubernetes/kubernetes/blob/3d6c5b2e98afaaae1d17107e2d3d709c726be49d/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L64-L68, so wouldn't actually increase load on *etcd*.

Thanks for the feedback though - I appreciate this is asking for something we definitely don't offer today and there's nuances to all approaches 😅

**Comment by serathius** [comment]
Since I was already working on using watch progress notifications for keeping the watch cache up-to-date (https://github.com/kubernetes/enhancements/issues/2340), I figured it might be handy for the Kubernetes clients too, and wanted to see if there was interest here first.

It's basically the same idea, just with different names ("bookmarks" in Kubernetes, "watch progress notification" in etcd). When working on the KEP we discovered many issues with etcd progress notifications, and I think the lessons learned could be useful for Kubernetes. For example, etcd's progress notifications aren't guaranteed. This would help clients keep their caches more up-to-date and reduce the risk of inconsistencies.

For consisted reads from watch cache we were able to work around those issues, however the lessons learned and ideas from etcd could also be useful for bringing this feature to Kubernetes. What do you think about working together to make a really solid watch notification system that works well for both Kubernetes and etcd? I think it would make caching way more consistent and reliable across the board.

**Comment by munnerz** [comment]
Absolutely, I'd love to work on improving any issues identified in the current watch notification system. That said, I'm still keen to see how (or if) we can expose this to kube-apiserver clients as well - do some of the issues we've found in etcd's handling of these notifications also impact how we might want to expose this to k8s users (either direction of implementation, or scalability/deliverability guarantees?).

Not being guaranteed is something that definitely needs to be investigated as this will cause requests to timeout, but the same is true for the apiserver today, and getting some form of parity with the apiserver's behaviour within a 'front proxy' in front of the apiserver would be a big win for apiserver extensibility *out* of tree/core k8s :)

**Comment by serathius** [comment]
/retitle Allowing requesting watch bookmarks for consistent read from informer

**Comment by serathius** [comment]
/retitle Allow requesting watch bookmarks for consistent read from informer

**Comment by munnerz** [comment]
> /retitle Allow requesting watch bookmarks for consistent read from informer

How do you feel about adding this feature? If we did, exposing the bookmarks via the reflector should allow us to extend SharedIndexInformer to actually permit consistent reads, which could be a great way to ensure clients that today utilise live reads to ensure correctness, to reduce the apiserver burden down to that of just periodically sending a bookmark.

I think the biggest blocker/question in all this is:

* are there scalability concerns sending these bookmarks, and if so:
  1) how can clients control the bookmark frequency (if at all)
  2) how can administrators control minimum/maximum frequencies that clients can request

I am pretty convinced that we should utilise a time based approach rather than on-demand, as the use cases for this are likely quite specialised (and could be exposed with a "ConsistentReadWaitDuration" option on the informer), as this also avoids adding a network roundtrip to the apiserver (precisely the sort of request we are aiming to reduce).

In my case, I am using a reflector to ultimately feed a proxy with a watch cache, so being able to request that a bookmark be sent every ~100ms (or something akin/compatible with the existing watch cache bookmark durations) would allow me to serve consistent reads to my own clients.

Could we have a global administrator controlled limit on min/max watch bookmark durations (or a global default/fixed value we can tune with a flag), as well something like an RBAC verb evaluated at the time of establishing the watch to determine if a client is allowed to request periodic bookmarks for a given resource type? eg:

```
kind: Role
metadata:
  namespace: abc
rules:
- apiGroups: [""]
  resources: ["pods"]
  verbs: ["watchprogress"]
```

Being able to restrict to a subset of resource types (and also restrict generally with RBAC) seems like it make this less scary if we are worried about all informers one day getting bookmarks every 100ms due to some controller bug.

**Comment by munnerz** [comment]
I believe this would also help with use cases like Karmada's aggregated-apiserver: https://karmada.io/docs/reference/components/karmada-aggregated-apiserver

This instantiates a watch cache (from k/apiserver) per remote store/endpoint, however they will presumably be unable to safely handle consistent reads due to the lack of regular bookmarks on the wire.

**Comment by munnerz** [comment]
Or am I getting something wrong in how bookmarking works in the watch cache, requiring the request to actually be instantiated explicitly by a client to function in order to maintain correctness? 👀

**Comment by liggitt** [comment]
> How do you feel about adding this feature?
> ...
> In my case, I am using a reflector to ultimately feed a proxy with a watch cache, so being able to request that a bookmark be sent every ~100ms

The fan out effect of having the API server send bookmark events out to N watch clients in steady state many times per second in an ~idle cluster seems pretty bad to me.
